File: src/axiom/frontends/tui/widgets/commands.py
"""Slash commands — registry data plus the autocomplete menu widget.

The command list is data (name, usage, description); the actions themselves live
in :mod:`axiom.frontends.tui.app`, where they are wired to real core calls.
"""

from __future__ import annotations

from dataclasses import dataclass

from textual.containers import Vertical
from textual.message import Message
from textual.widgets import OptionList, Static
from textual.widgets.option_list import Option

from axiom.shared import theme


@dataclass(frozen=True)
class Command:
    """One slash command."""

    name: str
    description: str
    argument_hint: str = ""

    @property
    def usage(self) -> str:
        return f"{self.name} {self.argument_hint}".strip()


#: Canonical command set (OpenCode-style).
COMMANDS: tuple[Command, ...] = (
    Command("/help", "Show help"),
    Command("/model", "Change model", "[name]"),
    Command("/models", "Model list"),
    Command("/clear", "Clear conversation"),
    Command("/history", "Conversation history"),
    Command("/new", "New conversation"),
    Command("/settings", "Settings"),
    Command("/search", "Web search", "query"),
    Command("/status", "System status"),
    Command("/permissions", "Permission mode (ask/auto_approve_safe/auto_approve_all)"),
    Command("/profiles", "List / switch system prompt profiles"),
    Command("/exit", "Exit Axiom"),
)

_COMMAND_INDEX = {command.name: command for command in COMMANDS}


def find_command(token: str) -> Command | None:
    """Resolve ``/mo`` style tokens to a real command (exact name first)."""
    token = token.strip().split()[0] if token.strip() else ""
    if token in _COMMAND_INDEX:
        return _COMMAND_INDEX[token]
    matches = matching_commands(token)
    if len(matches) == 1:
        return matches[0]
    return None


def matching_commands(prefix: str) -> list[Command]:
    """Filter commands by the text typed after ``/`` (fuzzy on name)."""
    prefix = prefix.lstrip("/").lower()
    if not prefix:
        return list(COMMANDS)
    exact = [c for c in COMMANDS if c.name[1:] == prefix]
    if exact:
        return exact
    starts = [c for c in COMMANDS if c.name[1:].startswith(prefix)]
    if starts:
        return starts
    return [c for c in COMMANDS if prefix in c.name[1:]]


def command_rows(commands: list[Command], width: int = 60) -> list[Option]:
    """Render commands as option rows, aligned for the menu panel."""
    widest = max((len(c.usage) for c in commands), default=8)
    widest = min(widest, max(10, width - 24))
    rows: list[Option] = []
    for command in commands:
        usage = command.usage.ljust(widest)
        rows.append(Option(f"{usage}  {command.description}", id=command.name))
    return rows


class MenuOptionList(OptionList):
    """An OptionList that never takes focus.

    The slash menu must not steal focus from the prompt: the menu is
    navigated by InputBar.on_key while the caret stays in the input.
    """

    can_focus = False


class CommandMenu(Vertical):
    """The ``/`` autocomplete panel — filtered, arrow-navigable, mouse-clickable."""

    class Chosen(Message):
        """A command was accepted (Enter, Tab or click)."""

        def __init__(self, command: Command) -> None:
            self.command = command
            super().__init__()

    def __init__(self, **kwargs) -> None:
        super().__init__(id="command-menu", **kwargs)
        self._visible = False

    def compose(self):
        yield Static("COMMANDS", id="command-menu-title")
        yield MenuOptionList(id="command-menu-list")


    @property
    def open(self) -> bool:
        return self._visible

    def on_mount(self) -> None:
        self.display = False

    def show_for(self, text: str) -> None:
        """Open the menu and filter it for the current input text."""
        if not text.startswith("/"):
            self.hide()
            return
        token = text.split(" ", 1)[0]
        # once a space is typed the command itself is settled: hide the menu
        # (text, not text.strip() — a trailing space starts the argument too)
        argument_typed = " " in text
        commands = matching_commands(token)
        if argument_typed:
            self.hide()
            return
        if not commands:
            self.hide()
            return
        option_list = self.query_one("#command-menu-list", OptionList)
        option_list.clear_options()
        for option in command_rows(commands, self.size.width or 60):
            option_list.add_option(option)
        option_list.highlighted = 0
        self._visible = True
        self.display = True
        self.refresh(layout=True)
        # The menu is part of the normal vertical flow, so focus must stay in
        # the prompt while it is open — do not touch focus here (arrows are
        # driven by InputBar.on_key, clicks arrive as Chosen messages).


    def hide(self) -> None:
        if not self._visible:
            return
        self._visible = False
        self.display = False
        self.refresh(layout=True)

    def highlighted_command(self) -> Command | None:
        option_list = self.query_one("#command-menu-list", OptionList)
        if option_list.highlighted is None:
            return None
        option = option_list.get_option_at_index(option_list.highlighted)
        return _COMMAND_INDEX.get(str(option.id)) if option.id else None

    def on_option_list_option_selected(self, event: OptionList.OptionSelected) -> None:
        command = _COMMAND_INDEX.get(str(event.option.id))
        if command is not None:
            self.post_message(self.Chosen(command))

    def border_color(self) -> str:
        return theme.GARNET

File: desktop/src/lib/commands.ts
/**
 * Slash commands — the GUI mirror of `axiom.frontends.tui.widgets.commands`.
 *
 * The list is data (name, argument hint, description); the actions live in
 * `useAxiom`, where every command is wired to a real core call. Nothing here is
 * decorative: a command that is listed always does something.
 */

export interface SlashCommand {
  name: string;
  description: string;
  argumentHint?: string;
  /** Hint shown in the palette for commands that open a panel. */
  group: "chat" | "models" | "workspace" | "system" | "harness";
}

export const COMMANDS: SlashCommand[] = [
  { name: "/help", description: "Справка по командам и горячим клавишам", group: "system" },
  { name: "/new", description: "Новый чат", group: "chat" },
  { name: "/clear", description: "Очистить текущий разговор", group: "chat" },
  { name: "/history", description: "История разговоров", group: "chat" },
  { name: "/model", description: "Переключить модель", argumentHint: "[name]", group: "models" },
  { name: "/models", description: "Модели, возможности, состояние", group: "models" },
  { name: "/permissions", description: "Режим разрешений: ask, auto_approve_safe, auto_approve_all", group: "harness" },
  { name: "/profiles", description: "Системные prompt-профили", group: "harness" },
  { name: "/trajectory", description: "Timeline запуска, tools, tokens и timing", group: "harness" },
  { name: "/providers", description: "Провайдеры, API-ключ, Test и выбор модели", group: "harness" },
  { name: "/agents", description: "Агенты, оркестратор и назначенные модели", group: "harness" },
  { name: "/orchestrate", description: "Запустить задачу через ANALYST, CODER, DEBUGGER, TESTER и REVIEWER", argumentHint: "задача", group: "harness" },
  { name: "/context", description: "Контекст: токены, сообщения, файлы", group: "chat" },
  { name: "/search", description: "Веб-поиск и ответ по источникам", argumentHint: "query", group: "workspace" },
  { name: "/tools", description: "Инструменты агента", group: "workspace" },
  { name: "/status", description: "Состояние Ollama, модели и ядра", group: "system" },
  { name: "/settings", description: "Настройки AXIOM", group: "system" },
  { name: "/exit", description: "Закрыть AXIOM", group: "system" },
];

export function matchingCommands(input: string): SlashCommand[] {
  const token = input.trim().split(" ")[0].toLowerCase();
  if (!token.startsWith("/")) return [];
  const prefix = token.slice(1);
  if (!prefix) return COMMANDS;
  const exact = COMMANDS.filter((c) => c.name.slice(1) === prefix);
  if (exact.length) return exact;
  const starts = COMMANDS.filter((c) => c.name.slice(1).startsWith(prefix));
  if (starts.length) return starts;
  return COMMANDS.filter((c) => c.name.slice(1).includes(prefix));
}

export function commandByName(name: string): SlashCommand | undefined {
  const clean = name.startsWith("/") ? name : `/${name}`;
  return COMMANDS.find((c) => c.name === clean);
}

/** Split "/search python 3.15" into command name + rest. */
export function parseCommand(input: string): { name: string; args: string } {
  const trimmed = input.trim();
  const space = trimmed.indexOf(" ");
  if (space === -1) return { name: trimmed.toLowerCase(), args: "" };
  return { name: trimmed.slice(0, space).toLowerCase(), args: trimmed.slice(space + 1).trim() };
}

export const SHORTCUTS: { keys: string; label: string }[] = [
  { keys: "Enter", label: "Отправить сообщение" },
  { keys: "Shift+Enter", label: "Перенос строки" },
  { keys: "Ctrl+Enter", label: "Отправить с принудительным веб-поиском" },
  { keys: "/", label: "Палитра команд (в пустом поле ввода)" },
  { keys: "Esc", label: "Остановить генерацию / закрыть панель" },
  { keys: "Ctrl+N", label: "Новый чат" },
  { keys: "Ctrl+B", label: "Показать/скрыть боковую панель" },
  { keys: "Ctrl+K", label: "Поиск по истории" },
  { keys: "Ctrl+,", label: "Настройки" },
  { keys: "Ctrl+/", label: "Фокус в поле ввода" },
  { keys: "↑", label: "В пустом поле — редактировать прошлое сообщение" },
  { keys: "Ctrl+C", label: "Копировать выделенный ответ" },
];
